src/jit/lower.cpp

```diff
@@ -1357,10 +1367,80 @@ void Lowering::LowerArg(GenTreeCall* call, GenTreePtr* ppArg)
         // If an extra node is returned, splice it in the right place in the tree.
         if (arg != putArg)
         {
-            ReplaceArgWithPutArgOrCopy(ppArg, putArg);
+            ReplaceArgWithPutArgOrBitcast(ppArg, putArg);
+        }
+    }
+}
+
+#ifdef _TARGET_ARMARCH_
+//------------------------------------------------------------------------
+// LowerFloatArg: Lower the float call argument on the arm platform.
+//
+// Arguments:
+//    arg    - The arg node
+//    info   - call argument info
+//    argNum - argument number
+//
+//
+// Return Value:
+//    Return nullptr, if no transformation was done;
+//    return arg if there was in place transformation;
+//    return a new tree if the root was changed.
+//
+GenTreePtr Lowering::LowerFloatArg(GenTreePtr arg, fgArgTabEntryPtr info, unsigned argNum)
+{
+    var_types type    = arg->TypeGet();
+    var_types intType = (type == TYP_DOUBLE) ? TYP_LONG : TYP_INT;
+
+    if (arg->OperIsFieldList())
+    {
+        assert(arg->isContained());
+        assert(argNum == 0);
+
+        unsigned fieldNum = 0;
+        for (GenTreeFieldList *list = arg->AsFieldList(); list != nullptr; list = list->Rest(), fieldNum++)
```

Comment:
I don't think this is correct for non-HFA struct types. I think that the caller of this method should always iterate through a FIELD_LIST, and then call `LowerFloatArg` on its list items that require it. I think that will also be cleaner.

Reply:
I do not understand, 

> the caller of this method should always iterate through a FIELD_LIST

we do not always have a FIELD_LIST . Do you want to move this if-condition to the `LowerArg`? Could you please give an example when this code is incorrect?

Reply:
What I meant was that *if* the arg is a `FIELD_LIST`, then you need to iterate through all of its members. Some may be floating point, and some may not.

Reply:
Got it, thank you. If it is possible than we have a `FIELD_LIST` with different types in it, which type is set on the `FIELD_LIST`? And how is register information encoded in `info->regnum`?

Reply:
Each `FIELD_LIST` should only have the type of the field. The presence of a `FIELD_LIST` node is an indication that the argument is an aggregate (either a struct or a decomposed long). When it is passed in register(s) the `PUTARG` node will be a multireg node that has the register list.

Reply:
`struct fgArgTabEntry:: regNumber regNum; // The (first) register to use when passing this argument, set to REG_STK for arguments passed on the stack.`

Because regnum is just an int, we can't encode that the first field list member is passed through int register, the second float is passed through float register and etc. There is no enough information to construct `PUTARG`.

Reply:
Hmm - then perhaps I misremember. It must be the case that on arm only HFAs are passed in registers (not other struct types). I still think that it would be better and cleaner to iterate through the `FIELD_LIST` in the caller to this method, and not have it call itself on the child of the `FIELD_LIST` node.

Reply:
On ARM, only HFA are passed in floating-point registers. All structs can be passed, or partially passed, in integer registers.

Reply:
> All structs can be passed, or partially passed, in integer registers.

So does that mean that if we have a `TYP_DOUBLE` field, if that struct is promoted, we could expect to encounter the `TYP_DOUBLE` in a `FIELD_LIST` under a `PUTARG_REG` or `PUTARG_SPLIT`?